### PR TITLE
feat(telemetry): add the ability to override span name with otel.name attribute

### LINKED ISSUE #5261: Keep router span name consistent and do not use operation name
## Is your feature request related to a problem? Please describe.
The span name for the service entry point for Apollo Router OpenTelemtry tracing is `<service_name>:<operation_name>`. This causes many issues with my dashboard linking in DataDog as I now have a high cardinality of spans.

Screenshot 2024-05-28 at 2 57 58 PM

Screenshot 2024-05-28 at 2 59 04 PM

## Describe the solution you'd like
It would be nice if instead we used just a standard name like `<serice_name>:router` or if this was configurable so I could set the span name myself. I would also like the operation name to be a configurable span attribute as I have to go to the supergraph to see this today. This is before the GraphQL parsing so what ever logic is being used to get the operation span name today should just become the attribute value

## Describe alternatives you've considered
I could maybe use a custom collector and transform all the spans

### PR DESCRIPTION
**Pull request by bnjjj**
It gives you the ability to override the span name by using custom telemetry with any [selectors](https://www.apollographql.com/docs/router/configuration/telemetry/instrumentation/selectors/) you want just by setting the `otel.name` attribute.

Example:

```yaml
telemetry:
  instrumentation:
    spans:
      router:
        otel.name:
           static: router # Override the span name to router 
```

Fixes #5261 

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [x] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: docs/source/configuration/telemetry/instrumentation/spans.mdx [hunk 1]
```diff
@@ -157,14 +157,20 @@ The `mode` option will be defaulted to `spec_compliant` in a future release, and
...
+Here is an example if you want to mark the `router` and `supergraph` span in error if you have a graphql error in the payload and you want to enforce the `router` span name to be `graphql_router`.
 
 ```yaml title="router.yaml"
 telemetry:
   instrumentation:
     spans:
       router:
         attributes:
+          otel.name:
+           static: graphql_router # Override the span name to graphql_router
```
Comment (BrynCooke) [inline docs/source/configuration/telemetry/instrumentation/spans.mdx:173]:
```suggestion
            static: graphql_router # Override the span name to graphql_router 
```

**Review by BrynCooke — APPROVED** [review]

File: docs/source/configuration/telemetry/instrumentation/spans.mdx [hunk 2]
```diff
@@ -157,14 +157,20 @@ The `mode` option will be defaulted to `spec_compliant` in a future release, and
 By default spans are marked in error only if the http status code is different than 200. If you want to mark a span in error for other reason you can override the `otel.status_code` attribute which is responsible to mark a span in error or not.
 If it's in error then `otel.status_code` = `error`, if not it will be `ok`.
 
-Here is an example if you want to mark the `router` and `supergraph` span in error if you have a graphql error in the payload.
+## Span name
```
Comment (abernix) [inline docs/source/configuration/telemetry/instrumentation/spans.mdx:160]: This page is about Spans, so maybe "Naming" is enough?

```suggestion
## Naming
```

File: docs/source/configuration/telemetry/instrumentation/spans.mdx [hunk 3]
```diff
@@ -157,14 +157,20 @@ The `mode` option will be defaulted to `spec_compliant` in a future release, and
...
+
```
Comment (abernix) [inline docs/source/configuration/telemetry/instrumentation/spans.mdx:161]: The section below this says how to change the names of the spans, but we don't describe what the default span name is (at least right here in this text).  Could we add some text here saying something like:

```
By default, we will use a span naming convention that aligns with the current [blah].

```

Thoughts?

File: apollo-router/src/plugins/telemetry/config_new/spans.rs [hunk 4]
```diff
@@ -556,6 +566,10 @@ mod test {
         assert!(values
             .iter()
             .any(|key_val| key_val.key == opentelemetry::Key::from_static_str("test")));
+
+        assert!(values.iter().any(|key_val| key_val.key
+            == opentelemetry::Key::from_static_str("otel.name")
+            && key_val.value == opentelemetry::Value::String(String::from("new_name").into())));
```
Comment (abernix) [inline apollo-router/src/plugins/telemetry/config_new/spans.rs:572]: Should there also be a test that asserts that when this isn't the case that the default names will be used?

File: .changesets/feat_bnjjj_feat_otel_name_attribute.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+### Add the ability to override span name with otel.name attribute ([Issue #5261](https://github.com/apollographql/router/issues/5261))
+
```
Comment (abernix) [inline .changesets/feat_bnjjj_feat_otel_name_attribute.md:2]: Thoughts? @shorgi 

```suggestion
### Override tracing span names using custom span selectors ([Issue #5261](https://github.com/apollographql/router/issues/5261))

```

File: apollo-router/src/plugins/telemetry/config_new/spans.rs (hunk 4 above)
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/config_new/spans.rs:572]: There are integration tests that already exist for that

**Comment by bnjjj** [comment]
@abernix I don't think it will entirely fix https://github.com/apollographql/router/issues/5282 so I won't include it here

File: docs/source/configuration/telemetry/instrumentation/spans.mdx (hunk 3 above)
Comment (bnjjj) [reply docs/source/configuration/telemetry/instrumentation/spans.mdx:164]: Added ! Thanks !

**Review by Geal — APPROVED** [review]

**Pull request merged**